src/scripts/main.js

```
@@ -1,3 +1,249 @@
 'use strict';
 
-// write code here
+class TableSorter {
+  /**
+   * Table selector
+   *
+   * @type {string}
+   */
+  static TABLE_SEL = 'table';
+
+  /**
+   * Selector for table head elements
+   *
+   * @type {string}
+   */
+  static TABLE_TH_SEL = ':scope > thead > tr > th';
+
+  /**
+   * Selector for table tbody
+   *
+   * @type {string}
+   */
+  static TABLE_TBODY_SEL = ':scope > tbody';
+
+  /**
+   * Selector for table rows
+   *
+   * @type {string}
+   */
+  static TABLE_TR_SEL = ':scope > tr';
+
+  /**
+   * Sorting state of column
+   *
+   * @type {{ASC: number, DESC: number, NONE: number}}
+   */
+  static STATE = {
+    ASC: 1,
+    DESC: -1,
+    NONE: 0,
+  };
+
+  /**
+   * Table element in HTML
+   * @type {HTMLElement}
+   */
+  _tableElement;
+
+  /**
+   * Sorter state
+   * @type Object
+   */
+  _sorterState = {
+    COLUMN_INDEX: -1,
+    SORT: TableSorter.STATE.NONE,
+  };
+
+  /**
+   * Custom comparator used for table
+   * @type Array
+   */
+  _customComparators = [];
+
+  /**
+   * Constructor
+   *
+   * @param {string} tableDomSelector
+   * @param {function[]} customComparators
+   * @return void
+   */
+  constructor(tableDomSelector, customComparators = []) {
+    if (typeof tableDomSelector !== 'string') {
+      throw new Error('Table DOM selector must be a string');
+    }
+
+    // table
+    this._tableElement = document.querySelector(tableDomSelector);
+
+    if (!Array.isArray(customComparators)) {
+      throw new Error('Custom comparators must be an array with comparators');
+    }
+
+    /**
+     * Process custom comparators and bind to object
+     */
+    for (let i = 0; i < customComparators.length; i++) {
+      this._customComparators[customComparators[i].index] =
+        customComparators[i].comparator;
+    }
```

You validate `customComparators` is an array which is fine, but when mapping them into `_customComparators` you directly use `customComparators[i].index` as an array key. This can create sparse arrays or accept invalid indexes. Consider validating that `index` is an integer >= 0 (and optionally within the number of header columns) before assignment, and maybe log/throw for invalid entries.

---

meh, another logic works with this